src/server/src/main.rs

```diff
@@ -1,23 +1,215 @@
+    let user_map = user_map.lock().unwrap();
+
+    if let Some(user) = user_map.get(&username) 
+    {
+        if let Some(session_id) = session_id
+        {
+            let session_map = session_map.lock().unwrap();
+            if let Some(sess_user) = session_map.get(session_id) 
+            {
+                // if current session matches. then return full data
+                println!("Session ID matched for user {}", username);
+                return HttpResponse::Ok().json(user);
+            } 
+        }
+        // else return public data 
+        let public_data = User 
+        {
+            username: user.username.clone(),
+            password: String::from("***"), 
+        };
+        HttpResponse::Ok().json(public_data)
+    } else 
+    {
+        HttpResponse::build(403).body("User does not exist.")
+    }
+}
+
+async fn update_user_handler(
+    req: HttpRequest,
+    path: web::types::Path<(String,)>,
+    update_value: web::types::Json<String>,
+    user_map: web::types::Data<UserMap>,
+    session_map: web::types::Data<SessionMap>,) -> impl Responder 
+{
+    let (username,) = path.into_inner();
+    let session_id = req.headers().get("SessionID").and_then(|val| val.to_str().ok());
+    let session_map = session_map.lock().unwrap();
+
+    if let Some(sess_user) = session_id.and_then(|sid| session_map.get(sid)) 
+    {
+        let mut user_map = user_map.lock().unwrap();
+        if let Some(user) = user_map.get_mut(&username) 
+        {
+            user.password = update_value.into_inner(); //whatever we want update to do, just pw rn
+            return HttpResponse::Ok().json(user); 
+        }
+    }
+    HttpResponse::build(403).body("Incorrect sessionId for user.")
+}
+
+async fn login_user_handler(
+    body: web::types::Json<User>,
+    path: web::types::Path<(String,)>,
+    user_map: web::types::Data<UserMap>,
+    session_map: web::types::Data<SessionMap>,) -> impl Responder 
+{
+    let login_user = body.into_inner();
+    let (username,) = path.into_inner();
+    let map = user_map.lock().unwrap();
+
+    if let Some(user) = map.get(&username) 
+    {
+        if user.password == login_user.password 
+        {
+            // generate new session ID -> spawns a child process for the user session -> stores the session ID and child pid
+            let session_id = Uuid::new_v4().to_string();
+            let session_process = spawn_user_session(username.clone());
+            let mut session_map = session_map.lock().unwrap();
+            session_map.insert(session_id.clone(), session_process);
+            return HttpResponse::Ok().json(session_id); //returns id token
+        }
+    }
+    HttpResponse::BadRequest().body("Invalid username/password combination.")
+}
+
+async fn logout_user_handler(
+    body: web::types::Json<String>,
+    path: web::types::Path<(String,)>,
+    session_map: web::types::Data<SessionMap>,) -> impl Responder 
+{
+    let session_id = body.into_inner();
```

Comment:
should probably be getting this from the header, since i imagine any API request from the client will already have it in there.

Reply:
@lupinclaw